TST: Fix indexes test inheritance / fixturization

TomAugspurger opened the pull request:
Closes https://github.com/pandas-dev/pandas/issues/23746

Before:

```
7355 passed, 852 skipped, 5 xfailed, 13 warnings in 75.20 seconds
```

After:

```
5740 passed, 724 skipped, 5 xfailed, 12 warnings in 67.73 seconds 
```

TomAugspurger committed:
TST: Fix test inheritence / fixturization

Closes https://github.com/pandas-dev/pandas/issues/23746

pep8speaks commented:
Hello @TomAugspurger! Thanks for submitting the PR.

 - There are no PEP8 issues in the file [`pandas/tests/indexes/common.py`](https://github.com/TomAugspurger/pandas/blob/7495d57ae969ffa8da5c679911db47f8919fabbc/pandas/tests/indexes/common.py) !

 - There are no PEP8 issues in the file [`pandas/tests/indexes/datetimelike.py`](https://github.com/TomAugspurger/pandas/blob/7495d57ae969ffa8da5c679911db47f8919fabbc/pandas/tests/indexes/datetimelike.py) !

 - There are no PEP8 issues in the file [`pandas/tests/indexes/test_base.py`](https://github.com/TomAugspurger/pandas/blob/7495d57ae969ffa8da5c679911db47f8919fabbc/pandas/tests/indexes/test_base.py) !

 - There are no PEP8 issues in the file [`pandas/tests/indexes/test_common.py`](https://github.com/TomAugspurger/pandas/blob/7495d57ae969ffa8da5c679911db47f8919fabbc/pandas/tests/indexes/test_common.py) !

 - There are no PEP8 issues in the file [`pandas/tests/indexes/test_numeric.py`](https://github.com/TomAugspurger/pandas/blob/7495d57ae969ffa8da5c679911db47f8919fabbc/pandas/tests/indexes/test_numeric.py) !

 - There are no PEP8 issues in the file [`pandas/tests/indexes/test_range.py`](https://github.com/TomAugspurger/pandas/blob/7495d57ae969ffa8da5c679911db47f8919fabbc/pandas/tests/indexes/test_range.py) !

jschendel reviewed (commented):
A couple comments not explicitly related to the diff, but in the spirit of the issue being resolved:

I think `test_unique_na` can be moved from `common.py` to `test_base.py`:  this test doesn't look to be taking advantage of inheritance at all, and appears to be exactly the same each time it's run by child classes inheriting from `Base`.

Both `test_has_duplicates` and `test_duplicated` in `common.py` are still using the `indices` fixture.  It doesn't look like they're causing an excess number of tests to be fully run, as these tests have a `pytest.skip` baked in to skip cases when `indices` doesn't match the type of index in the child class.  This would be causing an excess of tests being skipped though, and is kind of a gross pattern.  Would be fine leaving this as a follow up if that'd be preferred; seems like cleaning up these two tests could be in "good first issue" territory (or maybe a "good second issue" due to some complexity being present).

TomAugspurger committed:
test_unique_na -> base

TomAugspurger committed:
deduplicate test_duplicated

TomAugspurger commented:
> I think test_unique_na can be moved from common.py to test_base.py

done.

> Both test_has_duplicates and test_duplicated in common.py are still using the indices fixture.

hmm yes. As soon as I saw them using `self._holder`, I moved on from those tests. But those are indeed just skipping. I fixed this in aba63dae55c406a7bf5d2ce4dd36e206a03af9dc, if you could double check my work. I *think* that `type(indices)` is just fine for all these tests.

```
$ pytest pandas/tests/indexes/test_common.py -k test_duplicated -v
=================================================================== test session starts ====================================================================
platform darwin -- Python 3.7.0, pytest-3.10.0, py-1.6.0, pluggy-0.7.1 -- /Users/taugspurger/Envs/pandas-dev/bin/python3
cachedir: .pytest_cache
rootdir: /Users/taugspurger/sandbox/pandas, inifile: setup.cfg
plugins: xdist-1.23.2, repeat-0.7.0, forked-0.2, faulthandler-1.5.0, cov-2.6.0, hypothesis-3.59.1
collected 392 items / 350 deselected

pandas/tests/indexes/test_common.py::TestCommon::test_duplicated[Index0-first] PASSED                                                                [  2%]
pandas/tests/indexes/test_common.py::TestCommon::test_duplicated[Index0-last] PASSED                                                                 [  4%]
pandas/tests/indexes/test_common.py::TestCommon::test_duplicated[Index0-False] PASSED                                                                [  7%]
pandas/tests/indexes/test_common.py::TestCommon::test_duplicated[Index1-first] PASSED                                                                [  9%]
pandas/tests/indexes/test_common.py::TestCommon::test_duplicated[Index1-last] PASSED                                                                 [ 11%]
pandas/tests/indexes/test_common.py::TestCommon::test_duplicated[Index1-False] PASSED                                                                [ 14%]
pandas/tests/indexes/test_common.py::TestCommon::test_duplicated[DatetimeIndex-first] PASSED                                                         [ 16%]
pandas/tests/indexes/test_common.py::TestCommon::test_duplicated[DatetimeIndex-last] PASSED                                                          [ 19%]
pandas/tests/indexes/test_common.py::TestCommon::test_duplicated[DatetimeIndex-False] PASSED                                                         [ 21%]
pandas/tests/indexes/test_common.py::TestCommon::test_duplicated[PeriodIndex-first] PASSED                                                           [ 23%]
pandas/tests/indexes/test_common.py::TestCommon::test_duplicated[PeriodIndex-last] PASSED                                                            [ 26%]
pandas/tests/indexes/test_common.py::TestCommon::test_duplicated[PeriodIndex-False] PASSED                                                           [ 28%]
pandas/tests/indexes/test_common.py::TestCommon::test_duplicated[TimedeltaIndex-first] PASSED                                                        [ 30%]
pandas/tests/indexes/test_common.py::TestCommon::test_duplicated[TimedeltaIndex-last] PASSED                                                         [ 33%]
pandas/tests/indexes/test_common.py::TestCommon::test_duplicated[TimedeltaIndex-False] PASSED                                                        [ 35%]
pandas/tests/indexes/test_common.py::TestCommon::test_duplicated[Int64Index0-first] PASSED                                                           [ 38%]
pandas/tests/indexes/test_common.py::TestCommon::test_duplicated[Int64Index0-last] PASSED                                                            [ 40%]
pandas/tests/indexes/test_common.py::TestCommon::test_duplicated[Int64Index0-False] PASSED                                                           [ 42%]
pandas/tests/indexes/test_common.py::TestCommon::test_duplicated[UInt64Index-first] PASSED                                                           [ 45%]
pandas/tests/indexes/test_common.py::TestCommon::test_duplicated[UInt64Index-last] PASSED                                                            [ 47%]
pandas/tests/indexes/test_common.py::TestCommon::test_duplicated[UInt64Index-False] PASSED                                                           [ 50%]
pandas/tests/indexes/test_common.py::TestCommon::test_duplicated[RangeIndex-first] SKIPPED                                                           [ 52%]
pandas/tests/indexes/test_common.py::TestCommon::test_duplicated[RangeIndex-last] SKIPPED                                                            [ 54%]
pandas/tests/indexes/test_common.py::TestCommon::test_duplicated[RangeIndex-False] SKIPPED                                                           [ 57%]
pandas/tests/indexes/test_common.py::TestCommon::test_duplicated[Float64Index-first] PASSED                                                          [ 59%]
pandas/tests/indexes/test_common.py::TestCommon::test_duplicated[Float64Index-last] PASSED                                                           [ 61%]
pandas/tests/indexes/test_common.py::TestCommon::test_duplicated[Float64Index-False] PASSED                                                          [ 64%]
pandas/tests/indexes/test_common.py::TestCommon::test_duplicated[Index2-first] PASSED                                                                [ 66%]
pandas/tests/indexes/test_common.py::TestCommon::test_duplicated[Index2-last] PASSED                                                                 [ 69%]
pandas/tests/indexes/test_common.py::TestCommon::test_duplicated[Index2-False] PASSED                                                                [ 71%]
pandas/tests/indexes/test_common.py::TestCommon::test_duplicated[CategoricalIndex-first] PASSED                                                      [ 73%]
pandas/tests/indexes/test_common.py::TestCommon::test_duplicated[CategoricalIndex-last] PASSED                                                       [ 76%]
pandas/tests/indexes/test_common.py::TestCommon::test_duplicated[CategoricalIndex-False] PASSED                                                      [ 78%]
pandas/tests/indexes/test_common.py::TestCommon::test_duplicated[Index3-first] SKIPPED                                                               [ 80%]
pandas/tests/indexes/test_common.py::TestCommon::test_duplicated[Index3-last] SKIPPED                                                                [ 83%]
pandas/tests/indexes/test_common.py::TestCommon::test_duplicated[Index3-False] SKIPPED                                                               [ 85%]
pandas/tests/indexes/test_common.py::TestCommon::test_duplicated[MultiIndex-first] SKIPPED                                                           [ 88%]
pandas/tests/indexes/test_common.py::TestCommon::test_duplicated[MultiIndex-last] SKIPPED                                                            [ 90%]
pandas/tests/indexes/test_common.py::TestCommon::test_duplicated[MultiIndex-False] SKIPPED                                                           [ 92%]
pandas/tests/indexes/test_common.py::TestCommon::test_duplicated[Int64Index1-first] PASSED                                                           [ 95%]
pandas/tests/indexes/test_common.py::TestCommon::test_duplicated[Int64Index1-last] PASSED                                                            [ 97%]
pandas/tests/indexes/test_common.py::TestCommon::test_duplicated[Int64Index1-False] PASSED                                                           [100%]

```

jreback commented:
yeah this is ok. we need to finish moving to all fixtures from the inhertence setup of the original (might even be an open issue about this).

TomAugspurger committed:
skip rangeindex

jreback commented:
thanks
